functions/access.js

```diff
@@ -13,23 +13,30 @@ const checkIsValidToken = async (venueId, uid, token) => {
   return await admin
     .firestore()
     .runTransaction(async (transaction) => {
-      const venue = await transaction.get(venueRef);
-      const granted = await transaction.get(accessRef);
+      const [venue, granted] = await Promise.all([
+        transaction.get(venueRef),
+        transaction.get(accessRef),
+      ]);
 
-      if (!venue.exists) {
-        return false;
-      }
-      if (!granted.exists || !granted.tokens) {
+      if (!venue.exists || !granted.exists || !granted.tokens) {
         return false;
       }
 
       if (Object.keys(granted.tokens).includes(token)) {
         // Record that the token was checked
-        if (!granted.tokens[token].usedAt) {
-          granted.tokens[token].usedAt = [];
-        }
-        granted.tokens[token].usedAt.push(Date.now());
-        transaction.update(accessRef, granted);
+        const isTokenChecked = granted.tokens[token].usedAt;
+
+        const newToken = {
+          ...token,
+          usedAt: Date.now(),
+        };
+        // granted.tokens[token].usedAt.push(Date.now());
+        transaction.update(
+          accessRef,
+          admin.firestore.FieldValue.arrayUnion(
+            isTokenChecked ? token : newToken
+          )
+        );
```

Comment:
Have we checked that this works as expected? I'm not sure how `arrayUnion` will work with a complex object off the top of my head?

Reply:
This approach and `{merge:true}` work exactly the same, maybe we should keep it consistent and choose just one of them.

Reply:
I haven't deeply explored the nuances of `merge: true`. I feel like in most situations we probably want to check if an entry exists, then either create or update it explicitly (all within a transaction) rather than using `merge: true` (OOS for this PR, more just my thoughts on this pattern in future)

The `merge: true` case feels like it has the potential to accidentally (or maliciously) overwrite unintended keys if we don't carefully control all of the data fed to it, yet by being more explicit with `arrayUnion` like we are here, that risk seems far less likely.